general/development/tools/behat.md

```diff
@@ -0,0 +1,721 @@
+
+Note: This is usually an indicator that your development machine is not well tuned. A better option would be to find out where the bottleneck is. This is usually the database configuration.
+
+### New step definitions or features are not executed
+
+If you are adding new tests or steps definitions update the tests list
+
+```php
+php admin/tool/behat/cli/util.php --enable
+```
+
+** For parallel runs, all options for initialising parallel runs are valid **
+
+### Tests are failing
+
+<!-- cspell:ignore Bselenium -->
+If you followed all the steps and you receive an unknown weird error probably your browser version is not compatible with the Selenium version you are running. Please refer Working combinations to ensure you have correct [Acceptance testing/Browsers#Working combinations of OS.2BBrowser.2Bselenium](https://docs.moodle.org/dev/Acceptance_testing/Browsers#Working_combinations_of_OS.2BBrowser.2Bselenium) of them to run acceptance test.
+
+### The tests are failing, and the error message is completely useless
+
+For example, it just says "Error writing to database" with no stack trace.
+
+Add `-vv` command-line option to get very verbose output.
+
+### Errors during setup (before test are launched)
+
+Typical errors are:
+
+- Behat requirement not satisfied: http://127.0.0.1/m/stable_master is not available, ensure you specified correct url and that the server is set up and started.
+- `init.php` or `util.php` complain that `"Unknown error 1 This is not a behat test site!"`.  Delete the behat `wwwroot` (look in `config.php` for `$CFG->behat_dataroot`) and drop the behat DB tables (look in `config.php` for `$CFG->behat_prefix`).  Then try again.
+- Behat is configured but not enabled on this test site.
+In order to fix those errors please check that: the `behat_dataroot` has correct write permissions and that the `$CFG->behat*` variables are placed before the `lib/setup.php` include:
+
+```php
+ require_once(__DIR__ . '/lib/setup.php');
+```
+
+### Selenium server is not running
+
+#### Chrome specific
+
+If you are using chrome, you need to ensure that the driver matches the version of the installed chrome browser – which may change on OS updates/upgrades. Moodle or Selenium will not give the appropriate message – see [MDL-67659](https://tracker.moodle.org/browse/MDL-67659/). One solution is the one suggested in the issue and use Andrew Nicols' [Chromedriver Wrapper](https://github.com/andrewnicols/chromedriver-wrapper/) which will ensure you have the appropriate driver before running the tests.
+
+## External links
+
+- Vagrant profile with Moodle and Behat preconfigured: https://github.com/mackensen/moodle-hat
+- Docker containers for Moodle Developers and Behat: https://github.com/moodlehq/moodle-docker
+- Docker environment with Behat preconfigured : https://github.com/tobiga/docker_moodle_environment
```

Comment:
Why these External links haven't been moved to the "See also" section? I think it's easier having all them together in the same section :-)